src/analyses/goto_check.cpp

```diff
@@ -348,6 +349,39 @@ void goto_checkt::mod_by_zero_check(
     guard);
 }
 
+/// check a mod expression for the case INT_MIN % -1
+void goto_checkt::mod_overflow_check(const mod_exprt &expr, const guardt &guard)
+{
+  if(!enable_signed_overflow_check)
+    return;
+
+  const auto &type = expr.type();
+
+  if(type.id() == ID_signedbv)
+  {
+    // INT_MIN % -1 is, in principle, defined to be zero in
+    // ANSI C, C99, C++98, and C++11. Most compilers, however,
+    // fail to produce 0, and in some cases generate an exception.
+    // C11 explicitly makes this case undefined.
+
+    notequal_exprt int_min_neq(
+      expr.op0(), to_signedbv_type(type).smallest_expr());
+
+    notequal_exprt minus_one_neq(
+      expr.op1(), from_integer(-1, expr.op1().type()));
```

Comment:
What I still don't like very much are the number of further assumptions baked into this. The standard says that `a / b` must be representable. Only because of two's complement does this imply that this isn't representable exactly for the case of `b == -1`. I'm not sure we have ever specified that two's complement is the one and only representation for goto programs to use? Shouldn't we do what the standard says, and simply check for overflow of `a / b`?

Reply:
C11 and C++11 also allow one's complement and the signed magnitude representation. But then the test type.id() == ID_signedbv would fail (this is documented to be two's complement).

Note that the check for overflow of signed integer division is implemented in exactly the same way.

Reply:
My apologies for missing that `ID_signedbv` actually has this documented.